flink-runtime/src/main/java/org/apache/flink/runtime/query/QueryableStateClient.java

```diff
@@ -341,7 +341,34 @@ public void shutDown() {
 					return previous;
 				}
 			} else {
-				return cachedFuture;
+				// do not retain futures which failed as they will remain in
+				// the cache even if the error cause is not present any more
+				// and a new lookup may succeed
+				boolean isFailedFuture = false;
+				if (cachedFuture.isCompleted()) {
+					// find out if the future failed
+					try {
+						cachedFuture.value().get().get();
```

Comment:
I think you can do `cachedFuture.value().get().isFailure()` instead of catching the Exception here.